### PR TITLE
xgboost4j-spark cross validation

### PR DESCRIPTION
**Pull request by ebernhardson**
* Refactor XGBoost.scala to pull out common pieces of
  trainWithRDD and buildDistributedBoosters

* Add crossValidateRDD method to do cross validation on each
  partition and report the results

* Returns per-partition results. Up to users to decide what they
  want to do with this, parse out numbers and take means or whatever.

### CODE REVIEW

**Comment by CodingCat** [comment]
what's the point to have this instead of using spark ml's cv?

**Comment by ebernhardson** [comment]
It's a bit of a poor man's early stopping. I wanted the cv scores from each iteration so I can figure out at what point adding more trees is just over fitting the train data. 

Actual early stopping would be best, but that seemed much more involved.

**Comment by CodingCat** [comment]
Let's go to the target at once instead of merging some partial solutions and we are hard to remove it once it is there...

**Comment by ebernhardson** [comment]
I've looked over it a bit more, and I still think early stopping from spark is going to be much more of a pain than it's worth vs. a simplified CV implementation.

* Early stopping wouldn't be too hard to naively implement in xgboost4j itself, mostly a little tracking of evaluation results.

* As far as i can tell early stopping in a distributed system requires the same validation data to be available at all executors. Otherwise one executor would decide to quit while others want to continue. It's not clear from the python documentation or commits how early stopping with distributed training is handled. Perhaps something could be worked out for distributed validation sets with with Rabit.allReduce, but the xgboost4j implementation is commented explicitly as only meant for testing purposes.

* Early stopping requires multiple DMatrices, at a minimum a train and validation matrix. xgboost4j-spark currently has no support for multiple matrices, although something could of course be worked out. If the validation set isn't too big it could be broadcasted, but the validation set does need to be large enough to be representative and a few tens of MB might not cut it.

* Providing multiple matrices with spark's partitioning model isn't very obvious. All matrices need to be represented in a single Dataset, perhaps with a column indicating which matrix each belongs to. I didn't look too deeply but it doesn't seem that spark.ml has a concept of multi-way splits for test/train/validation. Doing this with an RDD would be awkward, requiring perhaps RDD[(MLLabeledPoint, Int)] or some such to indicate the matrix id of each point. Perhaps to have a reasonably clean implementation we would want to transition away from the trainWithRDD interface, instead making trainWithRDD call into trainWithDataFrame, rather than the other way around. trainWithDataFrame makes groupData a bit harder to reason about though, and I'm specifically interesting in ranking tasks.

* If we want to mimic the python version and have consistency between languages, should return a model with all trained trees along with some extra data about which tree was the best tree. The serialization model for the xgboost4j Booster doesn't really have anywhere for this though, as the format is strictly limited to the model returned by the native library.

None of this is impossible to work out, although I'm sure as work progressed towards an implementation more problems would surface. It is an incredible amount of complexity vs. the approach of using the builtin cross validation code to get 90% of the way there though.

**Comment by CodingCat** [comment]
You can use distributed validation dataset and use rabit allreduce...that's what we discussed before but just no bandwidth to implement it

As I said, I will not merge an partial implementation just because a more comprehensive one is feasible but needs more efforts...once merged in, we cannot remove

**Comment by CodingCat** [comment]
For this kind of contrbutions, we are planning to have a contrib directory holding it...still planning

**Comment by ebernhardson** [comment]
Any suggestions for how to accomplish the original goal, determine at what # of trees an xgboost4j-spark model is over fitting the training set?

**Comment by CodingCat** [comment]
is it what you are looking for? https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-example/src/main/scala/ml/dmlc/xgboost4j/scala/example/spark/SparkModelTuningTool.scala#L165

**Comment by ebernhardson** [comment]
ParamGrid isn't really the right approach here, it would be computationally very expensive to search in this manner. ParamGrid is essentially doing brute force grid search over the provided list. So as an example to determine the number of trees at eta 0.05 you would need to build a model at 100 trees, and at 110 trees, and at 120 trees, and at 130 trees, etc. up to some arbitrary point. But the largest model has all the information necessary to answer the question about 100 trees, 110 trees, 120 trees, without actually building them. Because they've already been built. The crossValidation api is essentially just reporting the evaluation set at all of these iterations, The actual crossValidation part of it isn't particularly important, that is just where the underlying xgboost4j library supports reporting per-iteration evaluation scores.

Perhaps some method could allow evaluating a test set, and return the evaluation result at all rounds? There is some support in xgboost4j for this, with the treeLimit parameter of predict. xgboost4j would need to accept a treeLimit to the evalSet methods, the JNI and c_api would also have to accept that treeLimit, and in learner.EvalOneIter the treeLimit would have to be accepted and passed through to learner.predictRaw which already accepts this value.
